cclient/api/extended/get_inputs.c

```diff
@@ -4,63 +4,130 @@
+  if (addr_opt.total == 0) {
+    addr_index = 0;
+    addr_index_end = (unsigned)-1;
+  } else {
+    addr_index = addr_opt.start;
+    addr_index_end = addr_opt.total;
+  }
+
+  for (; addr_index < addr_index_end; addr_index++) {
+    balances_req = get_balances_req_new();
+    balances_res = get_balances_res_new();
+    log_debug(client_extended_logger_id, "[%s:%d]\n", __func__, __LINE__);
+    if (!balances_req || !balances_res) {
+      ret_code = RC_OOM;
+      log_error(client_extended_logger_id, "%s create get balances request or response object failed: %s\n", __func__,
+                error_2_string(ret_code));
+      goto done;
+    }
+
+    addr = iota_sign_address_gen_flex_trits(seed, addr_index, addr_opt.security);
+    if (addr) {
+      ret_code = hash243_queue_push(&balances_req->addresses, addr);
+      if (ret_code) {
+        log_error(client_extended_logger_id, "%s:%d hash queue push failed: %s\n", __func__, __LINE__,
+                  error_2_string(ret_code));
+        goto done;
+      }
+    } else {
+      // gen address failed.
+      ret_code = RC_OOM;
+      log_error(client_extended_logger_id, "%s address generation failed: %s\n", __func__, error_2_string(ret_code));
+      goto done;
+    }
+    
+    // For request without `addr_opt.total`, end this loop if the current addr is an unused address
+    if (addr_opt.total == 0) {
+      ret_code = is_unused_address(serv, addr, &is_unused);
+      if (ret_code == RC_OK && is_unused) {
+        break;
+      }
+    }
+    
+    balances_req->threshold = 100;  // currently `100` should be used
+
+    ret_code = iota_client_get_balances(serv, balances_req, balances_res);
+    if (ret_code == RC_OK && addr) {
+      // expect balance value is in order.
+      if (out_inputs->total_balance < threshold) {
+        input_t tmp_input = {};
+        tmp_input.balance = get_balances_res_balances_at(balances_res, 0);
+        tmp_input.key_index = addr_index;
         tmp_input.security = addr_opt.security;
-        memcpy(tmp_input.address, q_iter->hash, FLEX_TRIT_SIZE_243);
+        memcpy(tmp_input.address, addr, FLEX_TRIT_SIZE_243);
         if ((ret_code = inputs_append(out_inputs, &tmp_input)) != RC_OK) {
           log_error(client_extended_logger_id, "%s get new address failed: %s\n", __func__, error_2_string(ret_code));
           goto done;
         }
+      } else {
+        break;
       }
-      counter++;
-    }
-    // check if balance is sufficient
-    if (out_inputs->total_balance < (int64_t)threshold) {
-      ret_code = RC_CCLIENT_INSUFFICIENT_BALANCE;
-      log_warning(client_extended_logger_id, "insufficient balance\n");
     }
+
+    get_balances_req_free(&balances_req);
+    get_balances_res_free(&balances_res);
+  }
+
+  // check if balance is sufficient
+  if (out_inputs->total_balance < threshold) {
+    ret_code = RC_CCLIENT_INSUFFICIENT_BALANCE;
+    log_warning(client_extended_logger_id, "insufficient balance\n");
   }
 
 done:
+  free(addr);
```

Comment:
memory leak.

Reply:
sorry I used `asan` and examine the code but I don't know which object cause memory leak.

Reply:
how did you examine this API call? the `addr` has a memory leak issue from the logic, but I can update you a sanitizer report later.

Reply:
```
==30356==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 324 byte(s) in 4 object(s) allocated from:
	#0 0x7f6e02487108 in calloc (/usr/lib/x86_64-linux-gnu/libasan.so.5+0xee108)
	#1 0x55d92ab43984 in iota_sign_address_gen_flex_trits common/helpers/sign.c:92
	#2 0x55d92a9fb86f in iota_client_get_inputs cclient/api/extended/get_inputs.c:76
	#3 0x55d92a9e9738 in example_get_inputs cclient/api/examples/example_get_inputs.c:25
	#4 0x55d92a9e4de7 in main cclient/api/examples/cclient_examples.c:88
	#5 0x7f6e01683b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
```

Reply:
oh I got the reason why I didn't get the memory leak issue during testing last time!